lib/vm/src/instance.rs

```diff
@@ -81,19 +81,19 @@ where
         gas_limit: u64,
     ) -> Self {
         set_gas(&mut wasmer_instance, gas_limit);
-        let res = Instance {
+        leave_storage(wasmer_instance.context(), Some(deps.storage));
+        Instance {
             wasmer_instance: wasmer_instance,
             api: deps.api,
             type_storage: PhantomData::<S> {},
-        };
-        res.leave_storage(Some(deps.storage));
-        res
+        }
     }
 
     /// Takes ownership of instance and decomposes it into its components.
     /// The components we want to preserve are returned, the rest is dropped.
-    pub fn recycle(instance: Self) -> (wasmer_runtime_core::Instance, A) {
-        (instance.wasmer_instance, instance.api)
+    pub fn recycle(instance: Self) -> (wasmer_runtime_core::Instance, A, Option<S>) {
```

Comment:
Very nice, yes, consume Self

Reply:
Yeah, `Self` the type, not the reference. I tried to do the same as an instance method but that does not work trivially.

Reply:
This is good. It means it "consumes" the object, so no one can use it afterwards.

I *think* this means the destructor is called at the end, so any objects owned by `instance` at the end of the method are freed